src/software/ai/navigator/navigator.h

```diff
@@ -1,42 +1,154 @@
+     *
+     * @param dribble The DribbleIntent to visit
+     */
+    void visit(const DribbleIntent &dribble_intent) override;
+
+    /**
+     * Visits a KickIntent to perform an operation.
+     *
+     * @param kick_intent The KickIntent to visit
+     */
+    void visit(const KickIntent &kick_intent) override;
+
     /**
-     * Given the state of the world and a list of Intents the robots want to perform,
-     * returns a list of Primitives that our Robots should perform in order to
-     * work towards achieving their Intent.
+     * Visits a MoveIntent to perform an operation.
      *
-     * @param world The current state of the world
-     * @param assignedIntents A list of Intents assigned to our friendly robots
-     * @return A list of Primitives to be run by our Robots in order to work towards
-     * achieving their Intents
+     * @param move_intent The MoveIntent to visit
      */
-    virtual std::vector<std::unique_ptr<Primitive>> getAssignedPrimitives(
-        const World &world,
-        const std::vector<std::unique_ptr<Intent>> &assignedIntents) = 0;
+    void visit(const MoveIntent &move_intent) override;
 
-    virtual ~Navigator() = default;
+    /**
+     * Visits a MoveSpinIntent to perform an operation.
+     *
+     * @param move_spin_intent The MoveSpinIntent to visit
+     */
+    void visit(const MoveSpinIntent &move_spin_intent) override;
+
+    /**
+     * Visits a PivotIntent to perform an operation.
+     *
+     * @param pivot_intent The PivotIntent to visit
+     */
+    void visit(const PivotIntent &pivot_intent) override;
+
+    /**
+     * Visits a StopIntent to perform an operation.
+     *
+     * @param stop_intent The StopIntent to visit
+     */
+    void visit(const StopIntent &stop_intent) override;
+
+   private:
+    /**
+     * Create an obstacle for the given avoid area, with a buffer such that the edge
+     * of the robot does not protrude into the area
+     *
+     * @param avoid_area The area to convert into an obstacle
+     *
+     * @return A obstacle representing the given area
+     */
+    std::optional<Obstacle> obstacleFromAvoidArea(AvoidArea avoid_area);
+
+    // This navigators knowledge / state of the world
+    World world;
+
+    // The current Primitive the navigator has created from an Intent.
+    // This variable is set by each `visit` function
+    std::unique_ptr<Primitive> current_primitive;
+
+    // The current Robot the navigator has navigated for from an Intent.
+    // This variable is set by each `visit` function
+    std::optional<Robot> current_robot;
+
+    // The current destination the navigator has navigated to from an Intent.
+    // This variable is set by each `visit` function
+    Point current_destination;
+
+    std::vector<Obstacle> velocity_obstacles;
+
+    std::vector<std::vector<Point>> planned_paths;
 
     /**
-     * Returns the paths planned for each robot by the navigator
+     * Creates a list of obstacles to avoid based on avoid areas,
```

Comment:
We don't have avoid areas anymore?

Same comment goes for the rest of this PR

Reply:
Same comment as above: MotionConstraints haven't been integrated yet